# Pandas and Scikit-learn

Pandas is a Python library that contains high-level data structures and manipulation tools designed for data analysis. Think of Pandas as a Python version of Excel. Scikit-learn, on the other hand, is an open-source machine learning library for Python.

While Scikit-learn does a lot of the heavy lifting, what's equally important is ensuring that raw data is processed in such a way that we are able to 'feed' it to Scikit-learn. Hence, the ability to manipulate raw data with Pandas makes it an indispensible part of our toolkit.

# Kaggle

Kaggle is the leading platform for data science competitions. Participants compete for cash prizes by submitting the best predictive model to problems posted on the competition website.

https://www.kaggle.com/competitions

We will be reviewing the data from the Kaggle Titanic competition. Our aim is to make predictions on whether or not specific passengers on the Titanic survived, based on characteristics such as age, sex and class.

# Section 1-0 - First Cut

We will start by splitting the data into a training set and a test set. Next we process the training data, at which point the data will be used to 'train' (or 'fit') our model. With the trained model, we apply it to the test data to make the predictions. Finally, we then compare our predictions against the 'ground truth' to see how well our model performed.

It is very common to encounter missing values in a data set. In this section, we will take the simplest (or perhaps, simplistic) approach of ignoring the whole row if any part of it contains an NaN value. We will build on this approach in later sections.

## Pandas - Extracting data

First, we load the training data from a .csv file. This is the similar to the data found on the Kaggle website: 

https://www.kaggle.com/c/titanic/data

In [39]:
import pandas as pd
import numpy as np

df = pd.read_csv('data.csv')

We review the size of the data.

In [40]:
df.shape

(891, 12)

We now split the data into an 80% training set and 20% test set.

In [41]:
df_train = df.iloc[:712, :]
df_test = df.iloc[712:, :]

## Pandas - Cleaning data

We review a selection of the data. 

In [42]:
df_train.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

We notice that the columns describe features of the Titanic passengers, such as age, sex, and class. Of particular interest is the column Survived, which describes whether or not the passenger survived. When training our model, what we are essentially doing is assessing how each feature impacts whether or not the passenger survived (or if the feature makes an impact at all).

**Exercise**:
- Write the code to review the tail-end section of the data. 

In [43]:
df_train[10:]

PassengerId  Survived  Pclass  \
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
30            31         0       1   
31            32         1       1   
32            33         1       3   
33            34         0       2   
34            35         0       1   
35            36         0       1   
36            37         1       3   
37            38         0       3   
38            39         0       3   
39            40         1       3   
..           ...       ...     ...   
682          683         0       3   
683          684         0       3   
684          685         0       2   
685          686         0       2   
686          687         0       3   
687          688         0       3   
688          689         0       3   
689          690         1       1   
690          691         1       1   
691          692         1       3   
692          693         1       3   
693          694         0       3   
694          695         0       1   
695          696         0       2   
696          697         0       3   
697          698         1       3   
698          699         0       1   
699          700         0       3   
700          701         1       1   
701          702         1       1   
702          703         0       3   
703          704         0       3   
704          705         0       3   
705          706         0       2   
706          707         1       2   
707          708         1       1   
708          709         1       1   
709          710         1       3   
710          711         1       1   
711          712         0       1   

                                                  Name     Sex   Age  SibSp  \
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN      0   
20                                Fynney, Mr. Joseph J    male  35.0      0   
21                               Beesley, Mr. Lawrence    male  34.0      0   
22                         McGowan, Miss. Anna "Annie"  female  15.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
24                       Palsson, Miss. Torborg Danira  female   8.0      3   
25   Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
26                             Emir, Mr. Farred Chehab    male   NaN      0   
27                      Fortune, Mr. Charles Alexander    male  19.0      3   
28                       O'Dwyer, Miss. Ellen "Nellie"  female   NaN      0   
29                                 Todoroff, Mr. Lalio    male   NaN

We observe that the columns Name, Ticket and Cabin are, for our current purposes, irrelevant. We proceed to remove them from our data set.

In [44]:
df_train = df_train.drop(['Name', 'Ticket', 'Cabin'], axis=1)

Next, we review the type of data in the columns, and their respective counts.

In [45]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 712 entries, 0 to 711
Data columns (total 9 columns):
PassengerId    712 non-null int64
Survived       712 non-null int64
Pclass         712 non-null int64
Sex            712 non-null object
Age            565 non-null float64
SibSp          712 non-null int64
Parch          712 non-null int64
Fare           712 non-null float64
Embarked       711 non-null object
dtypes: float64(2), int64(5), object(2)
memory usage: 50.1+ KB


We notice that the columns Age and Embarked have NaNs or missing values. As previously discussed, we take the approach of simply removing the rows with missing values.

In [46]:
df_train = df_train.dropna()

**Question**

- If you were to fill in the missing values, with what values would you fill them with? Why?

Scikit-learn only takes numerical arrays as inputs. As such, we would need to convert the categorical columns Sex and Embarked into numerical ones. We first review the range of values for the column Sex, and map the string values to numbers.

In [47]:
df_train['Sex'].unique()

array(['male', 'female'], dtype=object)

In [48]:
df_train['Sex'] = df_train['Sex'].map({'female':0, 'male':1})

Similarly for Embarked, we review the range of values and map the string values to a numerical value that represents where the passenger embarked from.

In [49]:
df_train['Embarked'].unique()

array(['S', 'C', 'Q'], dtype=object)

In [50]:
df_train['Embarked'] = df_train['Embarked'].map({'C':1, 'S':2, 'Q':3})

**Question**
- What problems might we encounter by mapping C, S, and Q in the column Embarked to the values 1, 2, and 3? In other words, what does the ordering imply? Does the same problem exist for the column Sex?

In our final review of our training data, we check that (1) there are no NaN values, and (2) all the values are in numerical form.

In [51]:
df_train.head(10)

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0             1         0       3    1  22.0      1      0   7.2500         2
1             2         1       1    0  38.0      1      0  71.2833         1
2             3         1       3    0  26.0      0      0   7.9250         2
3             4         1       1    0  35.0      1      0  53.1000         2
4             5         0       3    1  35.0      0      0   8.0500         2
6             7         0       1    1  54.0      0      0  51.8625         2
7             8         0       3    1   2.0      3      1  21.0750         2
8             9         1       3    0  27.0      0      2  11.1333         2
9            10         1       2    0  14.0      1      0  30.0708         1
10           11         1       3    0   4.0      1      1  16.7000         2

In [52]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 564 entries, 0 to 710
Data columns (total 9 columns):
PassengerId    564 non-null int64
Survived       564 non-null int64
Pclass         564 non-null int64
Sex            564 non-null int64
Age            564 non-null float64
SibSp          564 non-null int64
Parch          564 non-null int64
Fare           564 non-null float64
Embarked       564 non-null int64
dtypes: float64(2), int64(7)
memory usage: 44.1 KB


Finally, we convert the processed training data from a Pandas dataframe into a numerical (Numpy) array, and create a column from the outcomes of the training data.

In [53]:
X_train = df_train.iloc[:, 2:].values  # je n'ai plus de dataframe, je ne prends que les valeurs sans la structure -> head() ne marche pas. 
y_train = df_train['Survived']

In [54]:
#print(y_train.head(10))
#X_train.head(10)  : marche sans .values

## Scikit-learn - Training the model

In this section, we'll simply use the model as a black box. We'll review more sophisticated techniques in later sections.

In particular, we'll be using the Random Forest model. The intuition is as follows: each feature is reviewed to see how much impact it makes to the outcome. The most prominent feature is segmented into a 'branch'. A collection of branches is a 'tree'. The Random Forest model, broadly speaking, creates a 'forest' of trees and aggregates the results.

http://en.wikipedia.org/wiki/Random_forest

In [55]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators=100, random_state=0)

We use the processed training data to 'train' (or 'fit') our model.

In [56]:
model = model.fit(X_train, y_train)  #on entraine le model sur les données 

## Scikit-learn - Making predictions

We now review a selection of the test data.

In [57]:
df_test.head(10)

PassengerId  Survived  Pclass  \
712          713         1       1   
713          714         0       3   
714          715         0       2   
715          716         0       3   
716          717         1       1   
717          718         1       2   
718          719         0       3   
719          720         0       3   
720          721         1       2   
721          722         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
712                    Taylor, Mr. Elmer Zebley    male  48.0      1      0   
713                  Larsson, Mr. August Viktor    male  29.0      0      0   
714                       Greenberg, Mr. Samuel    male  52.0      0      0   
715  Soholt, Mr. Peter Andreas Lauritz Andersen    male  19.0      0      0   
716               Endres, Miss. Caroline Louise  female  38.0      0      0   
717         Troutt, Miss. Edwina Celia "Winnie"  female  27.0      0      0   
718                         McEvoy, Mr. Michael    male   NaN      0      0   
719                Johnson, Mr. Malkolm Joackim    male  33.0      0      0   
720           Harper, Miss. Annie Jessie "Nina"  female   6.0      0      1   
721                   Jensen, Mr. Svend Lauritz    male  17.0      1      0   

       Ticket      Fare  Cabin Embarked  
712     19996   52.0000   C126        S  
713      7545    9.4833    NaN        S  
714    250647   13.0000    NaN        S  
715    348124    7.6500  F G73        S  
716  PC 17757  227.5250    C45        C  
717     34218   10.5000   E101        S  
718     36568   15.5000    NaN        Q  
719    347062    7.7750    NaN        S  
720    248727   33.0000    NaN        S  
721    350048    7.0542    NaN        S

As before, we process the test data in a similar fashion to what we did to the training data.

In [58]:
df_test = df_test.drop(['Name', 'Ticket', 'Cabin'], axis=1)

df_test = df_test.dropna()

df_test['Sex'] = df_test['Sex'].map({'female': 0, 'male':1})
df_test['Embarked'] = df_test['Embarked'].map({'C':1, 'S':2, 'Q':3})

X_test = df_test.iloc[:, 2:]
y_test = df_test['Survived']

We now apply the trained model to the test data (omitting the column PassengerId) to produce an output of predictions.

In [60]:
y_prediction = model.predict(X_test)

## Evaluation

Comparing our predictions against the actual values gives us a list of 0s and 1s, and adding up the elements of the list gives us the number of correct predictions.

In [61]:
np.sum(y_prediction == y_test)

123

To get a sense of how good our prediction is, we calculate the model's accuracy by dividing the number of correct predictions by the length of the array of actual values.

In [62]:
np.sum(y_prediction == y_test) / float(len(y_test))

0.83108108108108103

Hence our predictions are 84% accurate. We now compare this against our best guess, by looking at the proportion of 0s and 1s.

In [63]:
np.sum(y_test) / float(len(y_test))

0.3918918918918919

Hence 39% of the passengers survived (with value 1) and 61% did not survive. If we were to guess that all the passengers did not survive, we would have a 61% accuracy. Hence our model gives an improvement of 23%!

In this section, we took the simplest approach of ignoring missing values. We look to build on this approach in Section 1-1.